server/handlers.go

```diff
@@ -998,6 +1023,84 @@ func (s *Server) writeAccessToken(w http.ResponseWriter, idToken, accessToken, r
 	w.Write(data)
 }
 
+// Handler for the /userinfo endpoint
+func (s *Server) handleUserInfo(w http.ResponseWriter, r *http.Request) {
+
+	token := ""
+	if r.Header.Get("Authorization") != "" {
+		// Extract token type and token value from Authorization header
+		authorizationSubstrs := strings.Split(r.Header.Get("Authorization"), " ")
+		token = authorizationSubstrs[1]
+	} else if r.URL.Query().Get("access_token") != "" {
+		token = r.URL.Query().Get("access_token")
+	} else {
+		s.logger.Error("No token found")
+		return
+	}
+
+	decryptedToken, err := s.Decrypt(token)
+	if err != nil {
+		s.logger.Errorf("Internal server error: %v", err)
+		s.renderError(w, http.StatusInternalServerError, "Internal server error.")
+		return
+	}
+
+	accessToken := new(internal.AccessToken)
+	if err := internal.Unmarshal(decryptedToken, accessToken); err != nil {
+		s.logger.Errorf("Failed to unmarshal access token: %v", err)
+	}
+
+	// Find the connector used in this auth flow
+	conn, err := s.getConnector(accessToken.ConnectorId)
+
+	if err != nil {
+		s.logger.Errorf("Failed to get connector with id %q : %v", accessToken.ConnectorId, err)
+		s.renderError(w, http.StatusBadRequest, "Requested resource does not exist")
+		return
+	}
+
+	// Create a flexible map initialized with the existing accessToken data
+	// If the connector does not implement a GetUserInfo, this map is
+	// returned as is
+	user := map[string]interface{}{
+		"sub":            accessToken.Subject,
+		"name":           accessToken.Username,
+		"email":          accessToken.Email,
+		"email_verified": accessToken.EmailVerified,
+		"groups":         accessToken.Groups,
+		"expiry":         accessToken.Expiry,
+	}
+
+	switch conn := conn.Connector.(type) {
+	// Delegate the responsibility of coming up with the user info
+	// to the connector if it is a UserInfoConnector
+	// e.g. oidc connector
+	case connector.UserInfoConnector:
+
+		connData := []byte(accessToken.ConnectorData)
+		err = conn.GetUserInfo(connData, &user)
+
+		if err != nil {
+			s.logger.Errorf("Failed to retrieve user info: %v", err)
```

Comment:
This is also log-and-ignore. Let's avoid that -- the call `conn.GetUserInfo(connData, &user)` could have done anything to `user`, like, have it filled half-way.

Let's do something like this
```go
extraInfo, err := conn.GetUserInfo(connData)
if err != nil {
    // log, return 500, etc -- but definitely `return`
}
for k, v := range extraInfo {
    user[k] = v
}
```
What do you think?

Reply:
I was thinking exactly the same while reading through this, mutating `user` inside `GetUserInfo` is bound to lead to a bug somewhere down the line 😂